text/0000-f16b.md

```diff
@@ -0,0 +1,166 @@
+
+- Add, Sub, Mul, Div, Rem, and the Assign variants
+- Neg
+- Copy and Clone
+- Display and Debug
+- Default
+- LowerExp and UpperExp
+- FromStr
+- PartialEq and PartialOrd
+- Sum and Product
+- All built-in methods common to the `f32` and `f64` types.
+
+The `f16b` type does not implement `From` for any integral type, as any such
+conversion could potentially overflow.
+
+`f32` and `f64` provide `impl From<f16b>`.
+
+A new module `std::f16b` will provide `f16b` versions of all the same constants
+as `std::f32` and `std::f64`, including their inner `consts` modules.
+
+The `f16b` type is FFI-safe, and may be used in foreign-function calls to
+C-compatible functions expecting a `bfloat16` value.
+
+Rust's `primitive_docs` will need an update to document the `f16b` type.
+
+A few external crates will need updates to support the new types,
+including `serde` and `num-traits`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds another floating-point type for developers to learn and select from,
+and increases the compiled size of the standard library (though the size
+observed by user code will remain the same if that code does not use the new
+functionality).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could use another name for `f16b`, such as `bfloat16` or `bf16`. The name
+`f16b` represents an attempt to align with existing floating-point types in
+Rust, and allow for other future types.
+
+We could make this type `f16`; however, there are two common 16-bit
+floating-point formats (`binary16` and `bfloat16`). This RFC does not preclude
+choosing one of those formats as `f16` in the future, but chooses to avoid
+pre-determining the answer to that question.
+
+We could support `f16b` exclusively on platforms with native hardware support.
+However, this would generate substantial conditional code within software
+wanting to use this type for memory-efficient storage and reduced memory
+bandwidth usage. Rather than forcing reimplementation of that conditional code,
+we can supply implementations on all platforms and allow code to use it
+uncondtionally. As precedent, note that Rust supports `i128` and `u128` on all
+platforms, whether or not those platforms have hardware support for 128-bit
+registers or 128-bit mathematical operations.
+
+We could support `bfloat16` via a separate crate, with no native support in
+Rust. However, native support would allow for native code generation (in LLVM
+or future backends), which a separate crate could not take advantage of. A
+separate crate could provide a fallback implementation via `f32` or `f64`, but
+not a native one.
+
+We should also provide hardware intrinsics for platforms with native `bfloat16`
+support. However, such intrinsics do not obviate the need for native code
+generation support in the compiler. Intrinsics only allow code to directly
+force the use of specific instructions, rather than supporting high-level
```

Comment:
I think @oli-obk may have covered this but I want to attach it to this line because it is wrong: intrinsics do not in general force the use of a particular instruction and in particular don't have to impede vectorization. Not only can Rust-level intrinsics expand to inline LLVM IR instruction sequences (e.g. as a radical example, `arch::x86_64::_mm256_setzero_ps` maps to the constant `zeroinitializer` in LLVM IR), even things that are intrinsics at the LLVM IR level can be detached from any particular target instruction and be as optimizable as anything else. For example, `@llvm.sqrt` can be constant folded, is subject to algebraic simplifications (tho most are only valid in fast-math mode), can be vectorized, etc. and any Rust program that calls `f{32,64}::sin` is using it.